extension/api.js

```diff
@@ -96,131 +283,25 @@ this.search = class extends ExtensionAPI {
       },
     });
 
-    function finish(result = false) {
-      Services.prefs.setBoolPref(RUN_ONCE_PREF, result);
-    }
-
-    // If the selected default is not the configured default for this region/locale, the
-    // user has changed default and we should not re-prompt.
-    let defaultEngine = await Services.search.getDefault();
-    let configuredDefault = await Services.search.originalDefaultEngine;
-    if (defaultEngine.name !== configuredDefault.name) {
-      console.log("reset-default-search: The default engine was already selected by the user.");
-      finish();
-      return;
-    }
+    // Ensure everything is started if this is an APP startup.
+    await searchInitialized;
+    await ExtensionSettingsStore.initialize();
 
-    // Get the latest installed addon that was installed prior to it's failure date.  Any non-app
-    // disabled state (ie. it is not disabled due to blocklisting) removes the addon from elibility.
-    let addons = (await AddonManager.getAddonsByIDs(
-        Array.from(lostEngines.keys())
-      )).filter(
-        a => a &&
-        !a.userDisabled &&
-        !a.softDisabled &&
-        !a.embedderDisabled &&
-        lostEngines.has(a.id) &&
-        a.installDate < lostEngines.get(a.id)
-      );
-    if (!addons.length) {
-      console.log("reset-default-search: No addons in our list are installed.");
-      finish();
+    await tryDefaultReset();
+    if (!tryagain) {
       return;
     }
 
-    // Filter out any that are blocklisted.
-    let enabledAddons = addons.filter(a => !a.appDisabled);
-
-    // If the extension is not running or enabled in the search service
-    // we will try again later.
-    let tryagain = false;
-
-    // We will only ask for the latest installed engine.  We will
-    // loop through the list until one of them makes it to the prompt.
-    enabledAddons.sort((a, b) => b.installDate - a.installDate);
-    for (let addon of enabledAddons) {
-      console.log(`reset-default-search: reset search engine to ${addon.id}`);
-
-      let policy = WebExtensionPolicy.getByID(addon.id);
-      if (!policy?.extension) {
-        tryagain = true;
-        console.log("reset-default-search: extension is not running, cannot set as default");
-        continue;
+    // Addons have not started when "update" is emitted, so we wait for "ready" to
+    // indicate the startup of an addon during the session.
+    Management.on("ready", async function tryUpdate(e, { id, resourceURI }) {
```

Comment:
When `ready` isn't triggered with the right condition prior to this extension unloading, then an event listener will be leaked

Reply:
The registration of the `Management.on("ready",` event is too late and can potentially miss registrations. That's because `tryDefaultReset` may still be running when add-ons of interest have been loaded (after the initial check).

Another issue with this listener is that it can result in multiple concurrent runs of `tryDefaultReset`, if there is a listener. To resolve this:

1. Add a listener for `"ready"` before calling `tryDefaultReset`.
2. Keep track of whether `tryDefaultReset` is pending (so that `tryDefaultReset` can potentially try again at the end of its execution). The "ready" listener should also be unregistered at the end of the `tryDefaultReset` call.
3. When the "ready" event is triggered, check if the flag has been set. If not, `tryDefaultReset` can be called. If not, set the flag so that `tryDefaultReset` can run again at the end.

The logic with `tryDefaultReset` currently assumes that the method does not throw. If you don't want to assume that, consider using a wrapper function that calls `finish` if needed, in a `try-await-catch-finally` block.

Reply:
I've cleaned some of this up based on other comments, but I'm not terribly concerned with startup/shutdown edge cases.  This will run again on next startup if the pref didn't get set.